Backbend Lifecycle

WhiredPlanck opened the pull request:
## Pull request

#### Issue tracker
Fixes will automatically close the related issues
<!-- Each issue should be on it's own line -->
Fixes # N/A

#### Feature
Describe features of this pull request

This PR is going to enhance the management of the librime backbend lifecycle, this is also a part of the improvment of the code quality. See commit records for details.

#### Code of conduct
- [x] [CONTRIBUTING](CONTRIBUTING.md)

#### Code style
- [x] `make sytle-lint`
- [x] [Conventional Commits](https://www.conventionalcommits.org/)

#### Build pass
- [x] `make debug`

#### Manually test
- [x] Done

#### Code Review
1. No wildcards import
2. Manual build and test pass
3. GitHub Action CI pass
4. At least one contributor review and approve
5. Merged clean without conflicts
6. PR will be merged by rebase upstream base

#### Daily build
Login and download artifact at https://github.com/osfans/trime/actions

#### Additional Info

WhiredPlanck committed:
refactor: remove redundant parameters to show composition window

WhiredPlanck committed:
refactor: adjust TextInputManager constructor method

Receives TrimeInputMethodService and RimeSeesion as parameters, instead of calling the static method to get their instances.

goofyz commented:
所以，這只是改善代碼，功能上沒有分別？

WhiredPlanck commented:
> 所以，這只是改善代碼，功能上沒有分別？

目前来说是的

WhiredPlanck committed:
refactor: introduce daemon to manage sessions access to rime

This will the successor of RimeWrapper

WhiredPlanck committed:
refactor: enhance lifecycle management of Rime

Migrate the abilities of RimeWrapper to RimeDaemon, the real Rime instance will be kept in RimeDaemon and shared via RimeSession

WhiredPlanck committed:
feat: show notification when restarting rime

This may prevent the rime backend from killing by the system.

WhiredPlanck commented:
@goofyz 现在新增了一个功能：在重新启动后端（部署）时，会显示一条通知，或许有助于保活 rime 后端不被系统杀掉。

goofyz commented:
切換到其他輸入法時，會出現 exception

```
java.lang.IllegalStateException: Trime not initialized
at com.osfans.trime.ime.core.TrimeInputMethodService$Companion.getService(TrimeInputMethodService.kt:1225)
at com.osfans.trime.data.db.DraftHelper.onInputEventChanged(DraftHelper.kt:73)
at com.osfans.trime.ime.core.TrimeInputMethodService$onFinishInputView$1.invoke(TrimeInputMethodService.kt:556)
at com.osfans.trime.ime.core.TrimeInputMethodService$onFinishInputView$1.invoke(TrimeInputMethodService.kt:554)
at com.osfans.trime.daemon.RimeDaemon$establish$1.runIfReady$lambda$3$lambda$2(RimeDaemon.kt:48)
at com.osfans.trime.daemon.RimeDaemon$establish$1.$r8$lambda$9LVEoyxIoiPIPHnt_T39dHQBFS4(Unknown Source:0)
at com.osfans.trime.daemon.RimeDaemon$establish$1$$ExternalSyntheticLambda0.run(Unknown Source:2)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7918)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

goofyz commented:
另外，rime 會在任何 UI  顯示前在一啟動時立刻 deploy。

這樣可以容易重現：
1. 殺掉 trime 程式
2. 刪除 `build` 資料夾
3. 點擊 text field (切換輸入法)  或 開啟「同文」設定頁面

若是點擊 text field 的話，它會在`InitializationUi` 顯示前已經在 deploying，結果是: 沒有反應，然後 deploy 完後才能顯示 `InitializationUi` 一秒左右再顯示 keyboard

若是開啟「同文」設定頁面，它會在 `SetupFragment` 還未顯示前 deploy ，要 deploy 完才能正常顯示頁面。

WhiredPlanck committed:
fix: crash on switching from other input method

The input method service instance might not be initialized yet, methods called on the service instance should judge the null state.

WhiredPlanck commented:
> 切換到其他輸入法時，會出現 exception

已在 https://github.com/osfans/trime/pull/1329/commits/e20d09d276e465231f08a3102794cfa963950f99 修复。

> 另外，rime 會在任何 UI 顯示前在一啟動時立刻 deploy。

> 若是開啟「同文」設定頁面，它會在 SetupFragment 還未顯示前 deploy ，要 deploy 完才能正常顯示頁面。

根据预期设计，为了能在应用设定界面顺畅调用 Rime 后端方法，在 ViewModel 中创建了一个公用的 RimeSession：

https://github.com/osfans/trime/blob/e20d09d276e465231f08a3102794cfa963950f99/app/src/main/java/com/osfans/trime/ui/main/MainViewModel.kt#L13

这个设计是参考自小企鹅的，小企鹅实现了后端独立线程异步运行，因此不会阻塞主线程（UI 线程）。但由于一些历史包袱，同文目前后端暂时只能也在主线程内运行（至少启停如此），这确实会影响 UI 渲染而降低使用体验。目前或许加个 `by lazy` 是较好的临时解决方案。

WhiredPlanck committed:
fix: create rime session by lazy in MainViewModel ...

... to prevent it slow down the UI rendering.

This is just a temporary workaround.

WhiredPlanck commented:
> 目前或许加个 `by lazy` 是较好的临时解决方案。

@goofyz https://github.com/osfans/trime/pull/1329/commits/ba7f0fab0aeadcb95ac1dc48b6ce3d6258293fe7

goofyz commented:
有點問題。
有時很正常，但切換輸入法時，有時會 endless loop 咁不斷重覆 deploy。
有時出現

```
Service com.osfans.trime.ime.core.TrimeInputMethodService has leaked IntentReceiver com.osfans.trime.ime.broadcast.IntentReceiver@a2f1875 that was originally registered here. Are you missing a call to unregisterReceiver()?
android.app.IntentReceiverLeaked: Service com.osfans.trime.ime.core.TrimeInputMethodService has leaked IntentReceiver com.osfans.trime.ime.broadcast.IntentReceiver@a2f1875 that was originally registered here. Are you missing a call to unregisterReceiver()?
	at android.app.LoadedApk$ReceiverDispatcher.<init>(LoadedApk.java:1828)
	at android.app.LoadedApk.getReceiverDispatcher(LoadedApk.java:1601)
	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1798)
	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1757)
	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:766)
	at androidx.core.content.ContextCompat$Api33Impl.registerReceiver(ContextCompat.java:1239)
	at androidx.core.content.ContextCompat.registerReceiver(ContextCompat.java:870)
	at androidx.core.content.ContextCompat.registerReceiver(ContextCompat.java:821)
	at com.osfans.trime.ime.broadcast.IntentReceiver.registerReceiver(IntentReceiver.kt:118)
	at com.osfans.trime.ime.text.TextInputManager.onCreate(TextInputManager.kt:107)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invoke(TrimeInputMethodService.kt:248)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invoke(TrimeInputMethodService.kt:240)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$1$1.invoke(RimeDaemon.kt:41)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$1$1.invoke(RimeDaemon.kt:41)
	at com.osfans.trime.core.RimeLifecycleKt.whenAtState$lambda$0(RimeLifecycle.kt:65)
	at com.osfans.trime.core.RimeLifecycleKt.$r8$lambda$I76GUho5NmdALNJsj6FFE5XRLzw(Unknown Source:0)
	at com.osfans.trime.core.RimeLifecycleKt$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7918)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```
有時又會

```
Process: com.osfans.trime.debug, PID: 29609
java.lang.NullPointerException: getCurrentInputEditorInfo(...) must not be null
	at com.osfans.trime.ime.core.TrimeInputMethodService$onColorChangeListener$1$1.invokeSuspend(TrimeInputMethodService.kt:124)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7918)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

WhiredPlanck committed:
fix: IntentReceiver was re-registered in TextInputManager

This may leak the receiver registered in TrimeInputMethodService, which leads to a potential crash.

WhiredPlanck committed:
fix: ensure to pass non-null input editor info to `startInput` of InputView ...

... on theme color changing to avoid a potential crash.

WhiredPlanck commented:
@goofyz https://github.com/osfans/trime/pull/1329/commits/af84b0995b69e9f16e6789a5ebbe03a7993de818 和 https://github.com/osfans/trime/pull/1329/commits/b7d66b06a2501d56b49a9be3c005c01218ca9b5f 已分别修好了这两个问题。

WhiredPlanck commented:
@goofyz @Bambooin If there are no other obvious issues, I plan to merge this PR tonight.

goofyz commented:
切換時始終會阻塞 UI 線程，也不算大問題。

WhiredPlanck commented:
> 切換時始終會阻塞 UI 線程，也不算大問題。

得在之后几轮 PR 中解决了，按照我目前的设想要实现的话会让修改的文件太多 ……
